src/treasury.cairo

```diff
@@ -171,23 +249,184 @@ mod Treasury {
+        ) -> Transfer {
             self.ownable.assert_only_owner();
             let token: IERC20Dispatcher = IERC20Dispatcher { contract_address: token_addr };
             assert(token.balanceOf(get_contract_address()) >= amount, Errors::INSUFFICIENT_FUNDS);
-            let status: bool = token.transfer(receiver, amount);
-            self.emit(TokenSent { receiver, token_addr, amount });
-            return status;
+            let transfers_count = self.transfers_count.read();
+            self.transfers_count.write(transfers_count + 1);
+            let transfer = Transfer {
+                id: transfers_count,
+                receiver,
+                token_addr,
+                amount,
+                cooldown_end: get_block_timestamp() + TREASURY_COOLDOWN_TIME,
+                status: TransferStatus::PENDING
+            };
+            self.transfers_on_cooldown.write(transfers_count, transfer);
+            self.emit(TransferPending { receiver, token_addr, amount });
+            self.transfers_on_cooldown.read(transfers_count)
+        }
+
+        fn cancel_transfer(ref self: ContractState, transfer_id: u64) {
+            assert(self.guardians.read(get_caller_address()), Errors::NOT_GUARDIAN);
+            assert(transfer_id < self.transfers_count.read(), Errors::INVALID_ID);
+            let initial_transfer = self.transfers_on_cooldown.read(transfer_id);
+            assert(
+                initial_transfer.status == TransferStatus::PENDING, Errors::TRANSFER_NOT_PENDING
+            );
+
+            let cancelation_event = TransferCancelled {
+                initial_amount: initial_transfer.amount,
+                token_addr: initial_transfer.token_addr,
+                receiver: initial_transfer.receiver
+            };
+            self
+                .transfers_on_cooldown
+                .write(
+                    transfer_id,
+                    Transfer { amount: 0, status: TransferStatus::CANCELLED, ..initial_transfer }
+                );
+            self.emit(cancelation_event);
+        }
+
+        fn execute_pending_by_id(ref self: ContractState, transfer_id: u64) -> bool {
+            assert(transfer_id < self.transfers_count.read(), Errors::INVALID_ID);
+            let transfer_pending = self.transfers_on_cooldown.read(transfer_id);
+            assert(
+                get_block_timestamp() >= transfer_pending.cooldown_end, Errors::COOLDOWN_NOT_PASSED
+            );
+
+            if transfer_pending.status == TransferStatus::CANCELLED {
+                return false;
+            }
+
+            let token: IERC20Dispatcher = IERC20Dispatcher {
+                contract_address: transfer_pending.token_addr
+            };
+            assert(
+                token.balanceOf(get_contract_address()) >= transfer_pending.amount,
+                Errors::INSUFFICIENT_FUNDS
+            );
+            let status: bool = token.transfer(transfer_pending.receiver, transfer_pending.amount);
```

Comment:
In line with Checks-Effects-Interactions pattern to prevent a reentrancy attack, please move the transfer to the end of the function (line 417)